src/NHibernate/Dialect/SQLiteDialect.cs

```diff
@@ -94,8 +102,34 @@ protected virtual void RegisterFunctions()
 
 			// NH-3787: SQLite requires the cast in SQL too for not defaulting to string.
 			RegisterFunction("transparentcast", new CastFunction());
-			
-			RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "substr(hex(?1), 7, 2) || substr(hex(?1), 5, 2) || substr(hex(?1), 3, 2) || substr(hex(?1), 1, 2) || '-' || substr(hex(?1), 11, 2) || substr(hex(?1), 9, 2) || '-' || substr(hex(?1), 15, 2) || substr(hex(?1), 13, 2) || '-' || substr(hex(?1), 17, 4) || '-' || substr(hex(?1), 21) "));
+
+			if (_binaryGuid)
+				RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "substr(hex(?1), 7, 2) || substr(hex(?1), 5, 2) || substr(hex(?1), 3, 2) || substr(hex(?1), 1, 2) || '-' || substr(hex(?1), 11, 2) || substr(hex(?1), 9, 2) || '-' || substr(hex(?1), 15, 2) || substr(hex(?1), 13, 2) || '-' || substr(hex(?1), 17, 4) || '-' || substr(hex(?1), 21) "));
+			else
+				RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "cast(?1 as char)"));
+		}
+
+
+		public override void Configure(IDictionary<string, string> settings)
```

Comment:
This is driver specific settings, not dialect's

Reply:
I'm afraid I'm not sure exactly what you mean. What are you proposing?

Reply:
What I'm saying is that BinaryGuid connection string option is specific to an ADO.NET provider implementation, not the dialect. For example, System.Data.SQLite and Mono.Data.Sqlite has this option, but Microsoft.Data.Sqlite does not. 

So, we need to handle this setting in a driver, not in a dialect.

Reply:
Ah, that's an interesting point. One thing to keep in mind though is that the proposed configurability defaults to assume that BinaryGuid=true, which is the only case supported by Microsoft.Data.Sqlite.

So the proposed handling in the dialect should not be a problem at runtime, even though handling it in the driver might be considered more correct code wise. At the moment I'm not sure how to resolve this in the driver, since it still has to affect what the dialect does.

Reply:
BinaryGuid connection string option is specific to the driver, but it still relates to how the database behaves about a specific type, which is a dialect business in my opinion.
That is just that some drivers do not support the non "binary guid" handling, so those drivers cannot be used with a database set to use non binary guid.
So I think this setting has to handled here in the dialect.